odl/discr/discr_mappings.py

```diff
@@ -687,42 +790,185 @@ def _evaluate(self, indices, norm_distances, out=None):
+        """Evaluate linear interpolation.
+
+        Modified for in-place evaluation and treatment of out-of-bounds
+        points by implicitly assuming 0 at the next node."""
         # slice for broadcasting over trailing dimensions in self.values
         vslice = (slice(None),) + (None,) * (self.values.ndim - len(indices))
 
-        # find relevant values
-        # each i and i+1 represents a edge
-        edges = product(*[[i, i + 1] for i in indices])
-        values = 0.
-        for edge_indices in edges:
-            weight = 1.
-            for ei, i, yi in zip(edge_indices, indices, norm_distances):
-                weight *= np.where(ei == i, 1 - yi, yi)
-            values += np.asarray(self.values[edge_indices]) * weight[vslice]
-        return values
+        out_shape = out_shape_from_meshgrid(norm_distances)
+        out_dtype = norm_distances[0].dtype
+
+        if out is None:
+            out = np.zeros(out_shape, dtype=out_dtype)
+        else:
+            out[:] = 0.0
+
+        # Weights and indices (per axis)
+        low_weights, high_weights, edge_indices = _create_lists(
+            indices, norm_distances, self.schemes, self.nn_variants)
+
+        # Iterate over all possible combinations of [i, i+1] for each
+        # axis, resulting in a loop of length 2**ndim
+        for lo_hi, edge in zip(product(*([['l', 'h']] * len(indices))),
+                               product(*edge_indices)):
+            weight = 1.0
+            for lh, w_lo, w_hi in zip(lo_hi, low_weights, high_weights):
+
+                # We don't multiply in place to exploit the cheap operations
+                # in the beginning: sizes grow gradually as following:
+                # (n, 1, 1, ...) -> (n, m, 1, ...) -> ...
+                # Hence, it is faster to build up the weight array instead
+                # of doing full-size operations from the beginning.
+                if lh == 'l':
+                    weight = weight * w_lo
+                else:
+                    weight = weight * w_hi
+            out += np.asarray(self.values[edge]) * weight[vslice]
+        return np.array(out, copy=False, ndmin=1)
 
 
-class _LinearMeshgridInterpolator(_MeshgridInterpolator,
-                                  _LinearPointwiseInterpolator):
+class _LinearInterpolator(_PerAxisInterpolator):
 
-    """Linear interpolator for point meshgrids."""
+    """Linear (i.e. bi-/tri-/multi-linear) interpolator.
+
+    Convenience class.
+    """
+
+    def __init__(self, coord_vecs, values, typ):
+        """Initialize a new instance.
+
+        coord_vecs : `tuple` of `numpy.ndarray`
+            Coordinate vectors defining the interpolation grid
+        values : array-like
+            Grid values to use for interpolation
+        typ : {'array', 'meshgrid'}
+            Type of expected input values in ``__call__``
+        """
+        super().__init__(coord_vecs, values, typ,
+                         schemes=['linear'] * len(coord_vecs),
```

Comment:
Same as above. The logic to allow for this is in the public operator, I deliberately kept this class as simple as possible.